### PR TITLE
Improve browser compatibility for Buffer usage

### LINKED ISSUE #28: Use Uint8Array instead of Buffer
Hey @pabigot, thanks for this library! We have been using it extensively for manual deserialization of blockchain data at Solana. I would like to propose that this library transition to be more browser friendly by using the `Uint8Array` type over the Node.js specific `Buffer` type. Currently we rely on browser polyfills but the latest bundler tools are dropping support for those so it'd be great to remove the need for polyfills entirely.

Are you open to this change being made to the next major release of `buffer-layout`? I'd be happy to own the transition since if we don't upstream this change, I'll be forking this package instead.

### PR DESCRIPTION
**Pull request by jstarry**
Fixes: https://github.com/pabigot/buffer-layout/issues/28

#### Changes
1. Explicitly require `buffer` module so that bundlers can polyfill it for browser apps but use the global `Buffer` if available.
2. Loosened type checks from `Buffer.isBuffer` to `instanceof Uint8Array` to improve API for browser apps

### CODE REVIEW

**Comment by jstarry** [comment]
@pabigot this is ready for review, thanks!

File: test/LayoutTest.js
Comment (pabigot) [inline test/LayoutTest.js]: Please follow existing style and use `it`.

File: lib/Layout.js
Comment (pabigot) [inline lib/Layout.js]: The [documentation](https://www.npmjs.com/package/buffer#usage) suggests this should be `... require('buffer/') ...` with a trailing slash.  Why isn't it?

File: package.json
Comment (pabigot) [inline package.json]: What motivates this change?

File: .travis.yml
Comment (pabigot) [inline .travis.yml]: Please combine this with the change to remove v4 support from package.json into a commit that explains why v4 can't be supported.  That commit should precede the changes to the implementation, since it's independent but (presumably) enables the main change.

File: lib/Layout.js
Comment (pabigot) [inline lib/Layout.js]: In the earlier commit you use:

    b = uint8ArrayToBuffer(b);

I'd rather see consistency so please change this to match.  Also `srcBuffer` should just change the variable `src`.

Also: why is this necessary?  The test seems to pass at the commit before this; if this is necessary there should be a new test case that fails without this change to demonstrate what's going on.

Depending on what motivated this change, it might be squashed into the one that changes all the other uses, so there bisecting wouldn't hit a commit that has a bug that resulted from the change in underlying implementation.

**Review by pabigot — CHANGES_REQUESTED** [review]
OK, this isn't quite what I thought.  Rather than replace the entry API use of `Buffer` with `Uint8Array`, it replaces the Node.js `Buffer` with a reimplementation.

I see this as being useful for browser applications, but a step backwards for Node.js applications.  The [performance](https://github.com/feross/buffer#node-01114) comparison suggests BrowserBuffer can be up to 4x slower than native Node.js (new, slice, and wrie operations particularly seem affected).

Can you come up with a solution that continues to use the native Buffer implementation in Node.js environments, but adds the external implementation only in browser environments?

If not, can you explain exactly why people who use this in Node.js won't be upset at the new unnecessary dependency and potential conflict of having multiple `Buffer` implementations in their applications?

If not, then does it make more sense for you to fork this project and publish your changes as `browser-buffer-layout`?  That'd also simplify ongoing maintenance here, as the only changes I've had to make in the last three years are to support an environment that wasn't a target for the original solution.

File: lib/Layout.js
Comment (jstarry) [reply lib/Layout.js]: > To require this module explicitly, use require('buffer/') which tells the node.js module lookup algorithm (also used by browserify) to use the npm module named buffer instead of the node.js core module named buffer!

Without the trailing slash, the node module resolution algorithm will choose the core buffer module

File: test/LayoutTest.js
Comment (jstarry) [reply test/LayoutTest.js]: These were the only uses of `it`, existing style uses `test(` everywhere else. This was causing linter errors

File: package.json
Comment (jstarry) [reply package.json]: Stated in the PR description: `Updated min node version from v4 to v5.10 to use Buffer.from instead of the deprecated new Buffer api`

Comment (jstarry) [reply package.json]: It wasn't necessary to change this as part of this PR, if you prefer more piecemeal changes, happy to split it up. But I don't think think new `buffer-layout` releases need this far back of node support. Folks using v4 can use old versions of `buffer-layout`

File: .travis.yml
Comment (jstarry) [reply .travis.yml]: It doesn't enable the main change, I will break this change up!

File: package.json
Comment (pabigot) [reply package.json]: OK.  Please put that sort of information into the commit messages or code comments.  Maintainers (me) should not have to go back to github to look at the PR to figure out what's going on.

File: lib/Layout.js
Comment (jstarry) [reply lib/Layout.js]: Motivation of this change is that if the API supports `Uint8Array` params, we need to polyfill the `Buffer.write` method which doesn't exist on `Uint8Array`.

File: package.json
Comment (jstarry) [reply package.json]: Sounds great, thanks for the feedback!

File: lib/Layout.js
Comment (pabigot) [reply lib/Layout.js]: Ah; so that addresses my concern about mixing Buffer implementations?  If so, great; since the upstream module doesn't explain that clearly, please add a comment in the code, and note in the commit message that this change doesn't affect uses in the Node.js environment (other than adding an unnecessary external dependency, I guess...).

Comment (jstarry) [reply lib/Layout.js]: Thanks for bringing up the extra external dependency, it's part of the reason I bumped the major version to v2. But I do agree it's not great. I will do some more exploring to figure out what the other viable alternatives would be.

File: lib/Layout.js
Comment (jstarry) [reply lib/Layout.js]: Added a new test case that fails without the conversion for `b` and `src`

File: lib/Layout.js
Comment (jstarry) [inline lib/Layout.js]: Without bumping the min node version, we need a little polyfill to account for the fact that node v4 doesn't have support for the `byteOffset` and `length` arguments to the `Buffer` constructor.

File: lib/Layout.js
Comment (jstarry) [reply lib/Layout.js]: Removed the new dependency since bundlers are smart enough to polyfill the `require('buffer').Buffer` as is. I've separated the `buffer` module changes into a new commit

**Comment by jstarry** [comment]
@pabigot I've taken your feedback into account and this PR is now ready for another pass, thanks!

File: test/LayoutTest.js
Comment (pabigot) [inline test/LayoutTest.js]: Name this one `basics`, since the effect of this patch series is to change the base packed data representation from `Buffer` to `Uint8Array`.  Then rename the old `basics` to `Buffer` to prove Buffer arguments still work.

File: CHANGELOG.md
Comment (pabigot) [inline CHANGELOG.md]: I intentionally referenced #26 because that was where the idea originated; I just chose to solve it in a more complete way.  I can see that might be confusing (though there's a comment in #26 pointing to #27), but still this doesn't belong in this commit.  Please drop it; I'll add a cleanup commit when I do the version update changes.

File: lib/Layout.js
Comment (pabigot) [reply lib/Layout.js]: As noted in the review summary, I'm going to remove v4 support so you won't need this.

**Review by pabigot — CHANGES_REQUESTED** [review]
This is getting close, thanks.

Overall I feel your change warrants a minor release update to 1.3.0. Because of that, I've also updated various things related to packages and integrations that hadn't been touched in the last three years. Those are now in the pu branch, which is queued to merge to the next branch.  Please take a look at #31, and comment on whether you have concerns.  After you've taken a look, or by the middle of next week, I'll merge that to the next branch.

Once those changes are in next I'd like you to retarget your PR to `next` instead of main.  Since #31 removes the support for Node 4 and provides a rationale for doing so, one of your polyfills won't be needed anymore.  Also I'd like you to remove release-related changes from your PR: the version number in package.json, and the description in CHANGELOG.md.  I'll handle those when I do the new release.  There will be one trivial conflict to resolve in `lib/Layout.js` when you rebase.

Also, please update all commit messages to the format:
```
   tag: short description

   details of change...

   Signed-off-by: yourname
```

The sentence/paragraph(s) in the body should describe the change and explain why it's being made, so somebody looking at changes in a commit can understand what they were intended to do.  For example for 4cd65bfff13d86a60a06a6e584d4a12ca46ede61 it should  be something like:
```
    Layout: Select Buffer implementation based on environment

    Use the presence of a variable defined in browser environments to
    select whether the Buffer implementation is built in to the
    environment or needs to be retrieved from a (potentially external)
    module.  This allows use of the package in Node.js, pure
    ECMAScript, and browser environments.

    Signed-off-by: Justin Starry <justin@solana.com>
```
ffb8f3a6d6986335764858dc1e3f297c9acc113f should explain why `Uint8Array` is preferred over Buffer, using text from #28.  And so on.

A good summary of the policy I've used is [here](http://www.openembedded.org/wiki/Commit_Patch_Message_Guidelines) or [here](https://www.kernel.org/doc/html/latest/process/submitting-patches.html).  The commits in #31 can also serve as an example.

**Comment by jstarry** [comment]
@pabigot I've prepped this branch so that it should be ready for merge once `pu` is merged. Really appreciate your feedback and diligence in maintainership!

File: CHANGELOG.md
Comment (pabigot) [reply CHANGELOG.md]: The CHANGELOG modification seems to still be present; I can remove it in next after merging or you can update your commit.  (Not sure why github thinks there are 13 commits in your PR; the targeting to next looks right, and the first 11 are already in that branch.  Weird.)

Also, could you do `npm run eslint` and fix the things it diagnoses?

Comment (jstarry) [reply CHANGELOG.md]: Fixed, sorry about the sloppiness!

File: lib/Layout.js
Comment (jstarry) [inline lib/Layout.js:137]: Ignored here, let me know if you disagree with taking this route

**Review by pabigot — APPROVED** [review]

**Comment by pabigot** [comment]
This looks fine.  Please confirm that this branch works as intended in a browser environment, since there's currently nothing in place to confirm that.

**Comment by jstarry** [comment]
Great, it shouldn't be too bad to add a test for that with mocha. I added an issue to track adding browser test coverage: https://github.com/pabigot/buffer-layout/issues/32

**Comment by jstarry** [comment]
Thanks again @pabigot, feel free to merge at your leisure, I don't have permission.

**Comment by pabigot** [comment]
This is a big change; I don't want to publish a version that claims to work in a browser without confirmation it works in that environment.  Can you install it in the environment you need it in and confirm that this specific code (not just your original version) works there?  I agree automated testing is highly desirable, but I don't even know how to do it manually.

**Comment by jstarry** [comment]
Sure thing, I have two branches here: https://github.com/jstarry/buffer-layout-test/branches. The `success` branch uses my changes and the `failure` branch uses the released v1.2.2 package.

**Pull request merged**
